File: TRPO/lab4/IntegerType.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;

public class IntegerType implements UserType {
    @Override
    public String typeName() {
        return "Integer";
    }

    @Override
    public Object create() {
        return 0;
    }

    @Override
    public Object clone(Object obj) {
        if (!(obj instanceof Integer)) 
            throw new IllegalArgumentException("Invalid object type");
        return Integer.valueOf((Integer) obj);
    }

    @Override
    public Object readValue(InputStreamReader in) throws IOException {
        BufferedReader reader = new BufferedReader(in);
        return parseValue(reader.readLine());
    }

    @Override
    public Object parseValue(String ss) {
        return Integer.parseInt(ss);
    }

    @Override
    public Comparator getTypeComparator() {
        return new Comparator() {
            @Override
            public int compare(Object o1, Object o2) {
                return Integer.compare((Integer) o1, (Integer) o2);
            }
        };
    }
}

File: TRPO/lab7/app/src/main/java/com/example/lab7/IntegerType.kt
package com.example.lab7

import java.io.InputStream

class IntegerType : UserType {
    override fun typeName() = "Integer"
    override fun create() = 0
    override fun clone(obj: Any) = obj as Int
    override fun readValue(input: InputStream) = input.bufferedReader().readLine().toInt()
    override fun parseValue(ss: String) = ss.toInt()
    override fun getComparator() = Comparator<Any> { o1, o2 -> (o1 as Int).compareTo(o2 as Int) }
}